src/python/pants/backend/scala/bsp/rules.py

```diff
@@ -0,0 +1,120 @@
+    ScalaBuildTarget,
+    ScalacOptionsItem,
+    ScalacOptionsParams,
+    ScalacOptionsResult,
+    ScalaPlatform,
+)
+from pants.backend.scala.dependency_inference.symbol_mapper import AllScalaTargets
+from pants.backend.scala.subsystems.scala import ScalaSubsystem
+from pants.base.build_root import BuildRoot
+from pants.bsp.rules import BSPBuildTargets, BSPBuildTargetsRequest
+from pants.bsp.spec import BuildTarget, BuildTargetCapabilities, BuildTargetIdentifier
+from pants.build_graph.address import AddressInput
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import QueryRule, collect_rules, rule
+from pants.engine.target import Target, WrappedTarget
+from pants.engine.unions import UnionRule
+from pants.jvm.subsystems import JvmSubsystem
+from pants.jvm.target_types import JvmResolveField
+
+
+class ScalaBSPBuildTargetsRequest(BSPBuildTargetsRequest):
+    pass
+
+
+def _pants_target_to_bsp_build_target(
+    target: Target, jvm: JvmSubsystem, scala: ScalaSubsystem
+) -> BuildTarget:
+    resolve = target[JvmResolveField].normalized_value(jvm)
+    scala_version = scala.version_for_resolve(resolve)
+    return BuildTarget(
+        id=BuildTargetIdentifier(uri=f"pants:{target.address}"),
+        display_name=str(target.address),
+        base_directory=None,
+        tags=(),
+        capabilities=BuildTargetCapabilities(),
+        language_ids=("scala",),
+        dependencies=(),
+        data_kind="scala",
+        data=ScalaBuildTarget(
+            scala_organization="unknown",
+            scala_version=".".join(scala_version.split(".")[0:2]),
+            scala_binary_version=scala_version,
+            platform=ScalaPlatform.JVM,
+            jars=(),
+        ),
+    )
+
+
+@rule
+async def determine_scala_bsp_build_targets(
+    _: ScalaBSPBuildTargetsRequest,
+    all_scala_targets: AllScalaTargets,
+    jvm: JvmSubsystem,
+    scala: ScalaSubsystem,
+) -> BSPBuildTargets:
+    scala_bsp_build_targets = [
+        _pants_target_to_bsp_build_target(tgt, jvm, scala) for tgt in all_scala_targets
+    ]
+    return BSPBuildTargets(targets=tuple(scala_bsp_build_targets))
+
+
+# -----------------------------------------------------------------------------------------------
+# Scalac Options Request
+# See https://build-server-protocol.github.io/docs/extensions/scala.html#scalac-options-request
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class HandleScalacOptionsRequest:
+    bsp_target_id: BuildTargetIdentifier
+
+
+@dataclass(frozen=True)
+class HandleScalacOptionsResult:
+    item: ScalacOptionsItem
+
+
+@rule
+async def handle_bsp_scalac_options_request(
+    request: HandleScalacOptionsRequest,
+    build_root: BuildRoot,
+) -> HandleScalacOptionsResult:
+    uri = request.bsp_target_id.uri
+    if not uri.startswith("pants:"):
+        raise ValueError(f"Unknown URI for Pants BSP: {uri}")
```

Comment:
Should this be validated in `BuildTargetIdentifier`? If it's a dataclsas, you can use `__post_init__`

Reply:
Yes since this code is duplicated. I'm thinking that this could be done in a rule that takes a `BuildTargetIdenifier` and returns a `Address`. Then the engine could convert via `await Get(WrappedTarget, BuildTargetIdentifier, request.bsp_target_id)`.

Reply:
I take it you can't put it on `BuildTargetIdentifier` because that type might be used in other contexts where it's fine to not start with `pants:`?

Another option might be using a method on it. I'm curious if we can do this through plain and simple Python, rather than more engine stuff which results in some indirection.

Reply:
The conversion code has to use the engine to convert the `AddressInput`, correct?

Reply:
Will clean this up in follow-on PR today.